### DISCUSSION TITLE
Injecting Navigator into ScreenModel

### DISCUSSION THREAD

**Discussion by KoTius** [adrielcafe/voyager#222] [Q&A] [answered]
I wonder if anything bad could possible happen if we simply pass the Navigator into ScreenModel?
Since all the methods of the Navigator are actually not composables (only `LocalNavigator.currentOrThrow` is) we can that way perfrom the navigation actions directly from the ViewModel without need to pass them back to UI and bother with LaunchEffect.

Sometimes we want to open different screens depend on data (is user is admin) or it could potentially help with implementation of the logout for 401 users.
Sometimes we want to open different screens depend on from which screen current is opened. E.g Search screens could open Item Detail screen and some times directly to Order Item screen.

```
class MyScreen : Screen {

    @Composable
    override fun Content() {

        val navigator = LocalNavigator.currentOrThrow

        val screenModel = rememberScreenModel { MyScreenModel(navigator) }
        
        MyScreenContent(screenModel)
    }
}
```

**Comment by Syer10** [comment, marked as answer]
You shouldn't do that, if the user puts the app in the background for a while, Jetpack Compose removes the UI so the Navigator won't survive, but the ScreenModel will. Doing this will make you always have a reference to the old navigator, and on restore will make it not work because its not the same navigator instance.

**Comment by Javi** [comment]
I'm trying out Voyager for the first time, and from experience having to navigate from views is much more mess-prone than navigating from viewmodels. I ran into the very problem @Syer10 mentioned, so I decided to 
```
            Navigator(ToBeDevelopedScreen()) {
                val mainScreenModel = it.getScreenModel<MainScreenModel>()
                mainScreenModel.navigator = it
            }
```
And then navigating from the model worked even after configuration changes and process death. The first problem I think would be figuring out a way to enforce this at compile time, as forgetting it would crash in runtime.

At the moment, I'll use this extension method and forbid the original one with something like Detekt maybe.
```
interface WithNavigator : ScreenModel {
    var navigator: Navigator
}

@Composable
inline fun <reified T : WithNavigator> Navigator.getScreenModel(): T =
    getNavigatorScreenModel<T> {
        parametersOf(this)
    }.apply { this.navigator = this@getScreenModel }
```
